apps/permissions.py

```diff
@@ -127,3 +143,45 @@ class IsOwner(permissions.BasePermission):
     def has_object_permission(self, request, view, obj):
         # Permissions are only allowed to the owner of the shipment.
         return has_owner_access(request, obj)
+
+
+class IsShipperCarrierModerator(permissions.BasePermission):
+    """
+    Custom permission to only allow owner, shipper, moderator and carrier
+    access to a shipment object on a nested route
+
+    This class should be used in conjunction with ShipmentExists
+    to validate the provided shipment_id in nested route
+
+    Example:
+        permission_classes = (IsAuthenticated, ShipmentExists, IsShipperCarrierModerator, )
+
+        On a ConfigurableGenericViewSet:
+
+        permission_classes = (IsAuthenticated, ShipmentExists, HasViewSetActionPermissions, IsShipperCarrierModerator, )
+
+    The declaration order of these classes matter.
+
+    """
+
+    def has_object_permission(self, request, view, obj):
+        return check_has_shipment_owner_access(request, obj.shipment, allowed_methods=None)
+
+    def has_permission(self, request, view):
+        shipment = Shipment.objects.get(id=view.kwargs.get('shipment_pk'))
+
+        return check_has_shipment_owner_access(request, shipment, allowed_methods=None)
+
+
+class ShipmentExists(permissions.BasePermission):
+
+    def has_permission(self, request, view):
+
+        shipment_id = view.kwargs.get('shipment_pk', None)
+
+        if not shipment_exists(shipment_id):
+            # The requested views are only accessible via nested routes
+            LOG.warning(f'User: {request.user}, is trying to access a non existing shipment: {shipment_id}')
```

Comment:
```suggestion
            LOG.warning(f'User: {request.user.id}, is trying to access a non-existent shipment: {shipment_id}')
```
Does request.user exist when profiles is disabled?